tools/pmd_test.bzl

```diff
@@ -7,38 +7,71 @@ def _impl(ctx):
         fail("Expecting a single java library")
 
     src_jar = lib.source_jars[0]
-    full_compile_jars = ":".join([f.path for f in lib.full_compile_jars.to_list()])
+    jar_deps = lib.full_compile_jars
+    full_compile_jars = ":".join([f.short_path for f in jar_deps.to_list()])
 
     # Presumably this needs to be updated to actually run the resulting shell script.
-    cmd = "@//tools:pmd"
 
-    args = [
-        cmd,
+    pmd_exe_file = ctx.attr._pmd[DefaultInfo].files_to_run.executable
+    ruleset = ctx.file.ruleset
+    pmd_cmd_args = [
+        pmd_exe_file.short_path,
         "-f text",
-        "-R pmd-ruleset.xml",
+        "-R {}".format(ruleset.short_path),
         "-d {}".format(src_jar.short_path),
-	"-auxclasspath {}".format(full_compile_jars),
+        "-auxclasspath {}".format(full_compile_jars),
+    ]
+    pmd_cmd = " ".join(pmd_cmd_args)
+    script = [
+        "#!/usr/bin/env bash",
+        "set -o errexit",
+        "set -o xtrace",
+        pmd_cmd,
     ]
-    argline = " ".join(args)
 
-    script = '\n'.join([
-        argline,
-    ])
-    print(script)
+    script_content = "\n".join(script)
+    print(script_content)
 
     # Write the file, it is executed by 'bazel test'.
+    test_file = ctx.actions.declare_file(ctx.attr.name)
     ctx.actions.write(
-        output = ctx.outputs.executable,
-        content = script,
+        output = test_file,
+        content = script_content,
+        is_executable = True,
     )
 
-    runfiles = ctx.runfiles(files=ctx.files.lib)
-    return [DefaultInfo(runfiles=runfiles)]
+    # Ensure everything that is needed to run PMD is in the runfiles tree:
+    pmd_runfiles = ctx.attr._pmd[DefaultInfo].data_runfiles.files
+    runfiles_jar_deps = depset(transitive = [
+        # Create a new depset w/ .to_list() to prevent a Java exception from being thrown in
+        # bazel 0.29.1 when combining depsets with different orders (in runfiles contruction).
+        depset(direct = jar_deps.to_list()),
+        pmd_runfiles,
+    ])
+    exe_runfile_deps = [ruleset, src_jar]
+    runfiles = ctx.runfiles(
+        transitive_files = runfiles_jar_deps,
+        files = exe_runfile_deps,
+    )
+    return [DefaultInfo(executable = test_file, runfiles = runfiles)]
```

Comment:
This part makes a lot of sense in retrospect, but it's not obvious to me how you got here. Was there any (public) documentation or examples you found particularly helpful in crafting this part?

[Sadly, I did not get to do much BUILD infra work internally, I was mostly in open source land]

Reply:
This is where I got the bits on setting the executable: https://docs.bazel.build/versions/master/skylark/rules.html#executable-rules-and-test-rules

If I were just reading that and trying to solve a "how do I make a test rule" I don't know if I'd succeed. Having a few other rule types (bundling/building) as a first step made making test rules for those build rules more manageable.

Reply:
Thanks again @AngusDavis!

Reply:
NP, @dhalperi - it was a fun excursion (also, I didn't know of PMD so its nice to have that in the toolbox).